In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# Loading data
customers = pd.read_csv('../data/raw/customers_dataset.csv')
sellers = pd.read_csv('../data/raw/sellers_dataset.csv')
geolocation = pd.read_csv('../data/raw/geolocation_dataset.csv')
order_reviews = pd.read_csv('../data/raw/order_reviews_dataset.csv')
order_items = pd.read_csv('../data/raw/order_items_dataset.csv')
orders = pd.read_csv('../data/raw/orders_dataset.csv')
products = pd.read_csv('../data/raw/products_dataset.csv')
product_category_name_translation = pd.read_csv('../data/raw/product_category_name_translation.csv')
order_payment = pd.read_csv('../data/raw/order_payments_dataset.csv')

## **Customers dataset**

In [3]:
customers.shape
print("My dataset has", customers.shape[0], "rows and", customers.shape[1], "columns")

My dataset has 99441 rows and 5 columns


In [4]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [5]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


Everything looks clean so far. We have no missing values and my data types are the way they should be.

In [6]:
# Conformation about the missing values
customers.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Now, it's time to check for unique values of customer city and customer state to check for spelling mistakes and typos  

In [7]:
customers['customer_city'].nunique()

4119

In [8]:
customers['customer_city'].value_counts().head(20)

customer_city
sao paulo                15540
rio de janeiro            6882
belo horizonte            2773
brasilia                  2131
curitiba                  1521
campinas                  1444
porto alegre              1379
salvador                  1245
guarulhos                 1189
sao bernardo do campo      938
niteroi                    849
santo andre                797
osasco                     746
santos                     713
goiania                    692
sao jose dos campos        691
fortaleza                  654
sorocaba                   633
recife                     613
florianopolis              570
Name: count, dtype: int64

In [9]:
sorted(customers['customer_city'].unique())

['abadia dos dourados',
 'abadiania',
 'abaete',
 'abaetetuba',
 'abaiara',
 'abaira',
 'abare',
 'abatia',
 'abdon batista',
 'abelardo luz',
 'abrantes',
 'abre campo',
 'abreu e lima',
 'acaiaca',
 'acailandia',
 'acajutiba',
 'acarau',
 'acari',
 'acegua',
 'acopiara',
 'acreuna',
 'acu',
 'acucena',
 'adamantina',
 'adhemar de barros',
 'adolfo',
 'adrianopolis',
 'adustina',
 'afogados da ingazeira',
 'afonso claudio',
 'afranio',
 'agisse',
 'agrestina',
 'agrolandia',
 'agronomica',
 'agua boa',
 'agua branca',
 'agua clara',
 'agua comprida',
 'agua doce',
 'agua doce do norte',
 'agua fria de goias',
 'agua limpa',
 'agua nova',
 'agua preta',
 'agua santa',
 'aguai',
 'aguas belas',
 'aguas claras',
 'aguas da prata',
 'aguas de lindoia',
 'aguas de santa barbara',
 'aguas de sao pedro',
 'aguas formosas',
 'aguas frias',
 'aguas lindas de goias',
 'aguas mornas',
 'aguas vermelhas',
 'agudo',
 'agudos',
 'aguia branca',
 'aimores',
 'aiuaba',
 'aiuruoca',
 'ajapi',
 'ajuric

We check for the values that have the lowest counts, because there is where its possible to be spelling mistakes

In [10]:
customers['customer_city'].value_counts().tail(10)

customer_city
machadinho                1
divino das laranjeiras    1
galia                     1
lagoa de sao francisco    1
uniao da serra            1
bequimao                  1
andarai                   1
vargem grande             1
curvelandia               1
eugenio de castro         1
Name: count, dtype: int64

In [11]:
# We now check for the customer state column for potential mistakes.
customers['customer_state'].nunique()

27

In [12]:
# Small number of unique values, so we can check them all
customers['customer_state'].value_counts()

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: int64

The customers dataset was assessed for missing values, duplicate records, primary key uniqueness, and inconsistencies in categorical variables. No significant data quality issues were identified. Based on these checks, the dataset is considered suitable for further analysis. Finally, I'm moving towards some final checks for the customers dataset.

In [13]:
# Checking for duplicated rows
customers.duplicated().sum()

0

In [14]:
customers['customer_id'].is_unique

True

This is expected and means that every customer_id identifies one customer record

In [15]:
customers["customer_id"].nunique()

99441

In [16]:
customers["customer_unique_id"].nunique()

96096

In [17]:
customers["customer_unique_id"].is_unique

False

This is crucial understanding the dataset. The dataset contains 96,096 unique customers, identified by customer_unique_id. The difference between the number of customer records (99,441) and unique customers indicates that some individuals placed more than one order during the observed period.

## **Sellers dataset**

In [18]:
sellers.shape

(3095, 4)

In [19]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


Very similar dataset with the previous one. We dont have a seller_unique_id here because it's obvious that its not needed. Below, I'm performing the same checks I did before.  

In [20]:
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


Again, there is no mistake like missing values here and my data types are the way they should be. Moving on

In [21]:
# Spelling mistakes
sellers["seller_city"].nunique()

611

In [22]:
sellers["seller_city"].value_counts().head()

seller_city
sao paulo         694
curitiba          127
rio de janeiro     96
belo horizonte     68
ribeirao preto     52
Name: count, dtype: int64

In [100]:
sellers['seller_city'].value_counts().tail(10)

seller_city
juzeiro do norte       1
angra dos reis         1
sp / sp                1
pirassununga           1
taruma                 1
s jose do rio preto    1
domingos martins       1
messias targino        1
ipe                    1
leme                   1
Name: count, dtype: int64

Here we do have something that we can fix. Obviously we should replace the last answer with just rio de janeiro.

In [24]:
sellers["seller_city"] = sellers["seller_city"].replace({
    "rio de janeiro, rio de janeiro, brasil": "rio de janeiro",
    "jacarei / sao paulo": "jacarei"
})

In [25]:
sellers[sellers["seller_city"] == "rio de janeiro"].value_counts().sum()

97

We see that the total value counts for rio de janeiro increased by one value in the column of seller city, which means that we replaced the inconsistent city naming successfully. The same goes with jacarei.

In [26]:
sellers["seller_state"].nunique()

23

In [27]:
sellers["seller_state"].value_counts()

seller_state
SP    1849
PR     349
MG     244
SC     190
RJ     171
RS     129
GO      40
DF      30
ES      23
BA      19
CE      13
PE       9
PB       6
RN       5
MS       5
MT       4
RO       2
SE       2
PI       1
AC       1
MA       1
AM       1
PA       1
Name: count, dtype: int64

In [28]:
# Checking for duplicated rows
sellers.duplicated().sum()

0

In [29]:
sellers['seller_id'].is_unique

True

## **Geolocation dataset**

In [30]:
geolocation.shape

(1000163, 5)

In [31]:
geolocation.head(10)

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP
5,1012,-23.547762,-46.635361,são paulo,SP
6,1047,-23.546273,-46.641225,sao paulo,SP
7,1013,-23.546923,-46.634264,sao paulo,SP
8,1029,-23.543769,-46.634278,sao paulo,SP
9,1011,-23.547640,-46.636032,sao paulo,SP


In [32]:
geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [33]:
geolocation["geolocation_city"].nunique()

8011

In [34]:
geolocation["geolocation_city"].value_counts().head(30)

geolocation_city
sao paulo                135800
rio de janeiro            62151
belo horizonte            27805
são paulo                 24918
curitiba                  16593
porto alegre              13521
salvador                  11865
guarulhos                 11340
brasilia                  10470
sao bernardo do campo      8112
osasco                     7658
santo andre                6863
niteroi                    6534
recife                     6168
goiania                    5661
fortaleza                  5538
campinas                   5479
sorocaba                   5361
santos                     5000
barueri                    4971
juiz de fora               4679
contagem                   4395
campo grande               4332
ribeirao preto             4187
florianopolis              4148
nova iguacu                4022
mogi das cruzes            3913
belem                      3789
sao jose dos campos        3759
sao goncalo                3601
Name: count, dtype: int

In [35]:
# The cities with the lowest value counts
geolocation["geolocation_city"].value_counts().tail()

geolocation_city
jacuípe                 1
mar vermelho            1
quebrangulo             1
poço das trincheiras    1
poxim                   1
Name: count, dtype: int64

In [36]:
# The total of value counts for every geolocation_city we want to search
geolocation[geolocation["geolocation_city"].str.contains("são paulo", case=False)].value_counts().sum()

24924

We observe, with the .head() method, that sao paulo has two spots in the top 4 value counts for geolocation_city. It is just written different. We have to change it because this can alter our outcome by a far.  

In [37]:
geolocation["geolocation_city"] = geolocation["geolocation_city"].replace({
    "são paulo": "sao paulo"
})

In [38]:
# Confirming that the change is done
geolocation["geolocation_city"].value_counts().head()

geolocation_city
sao paulo         160718
rio de janeiro     62151
belo horizonte     27805
curitiba           16593
porto alegre       13521
Name: count, dtype: int64

This is a pivotal point. I have spotted many spelling errors such as estação instead of estacao, or jordanésia instead of jordanesia and so many others. I will not focus on those spelling mistakes simply because they will not change my later results significantly much. More specifically:

In [39]:
print("The sum of value counts for the city estacao is equal to", geolocation[geolocation["geolocation_city"].str.contains("estacao", case=False)].value_counts().sum())


The sum of value counts for the city estacao is equal to 34


In [40]:
print("The sum of value counts for estação is equal to", geolocation[geolocation["geolocation_city"].str.contains("estação", case=False)].value_counts().sum())

The sum of value counts for estação is equal to 5


There is no reason for me to try to fix every single detail about my dataset, that i wont be needing in the future for further analysis. So we move on to the next dataset

In [41]:
# Lastly we check for duplicated rows
geolocation.duplicated().sum()

269682

Wow! Lets look deeper into this

In [42]:
duplicates = geolocation[geolocation.duplicated(keep=False)]
duplicates.sample(10, random_state=42)

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
114001,5025,-23.529233,-46.687040,sao paulo,SP
220781,9130,-23.689974,-46.503580,santo andre,SP
756960,65075,-2.508268,-44.303008,sao luis,MA
942382,90420,-30.036675,-51.201562,porto alegre,RS
915228,88535,-27.566755,-50.383686,correia pinto,SC
480781,24860,-22.784825,-42.917570,itaborai,RJ
239230,9850,-23.729229,-46.575110,sao bernardo do campo,SP
408177,20270,-22.916109,-43.221735,rio de janeiro,RJ
935513,89805,-27.088313,-52.615775,chapeco,SC
897685,87501,-23.761967,-53.307562,umuarama,PR


For this particular dataset duplicate values doesnt provide additional information, like transactions or order items. For this reason, the best way to handle this is to drop the above values

In [43]:
geolocation = geolocation.drop_duplicates()

In [44]:
# Confirmation
geolocation.duplicated().sum()

0

## **Order Reviews**

In [45]:
order_reviews.shape

(99224, 7)

In [46]:
order_reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


Its already obvious that there will be many NaN values here, especially int the fourth and fifth column. It is logical because not everyone will take the time to write about the product they bought

In [47]:
order_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


As expected. We have multiple NaN values in the fourth and fifth column. But, we have none in the other columns. With that being said, NaN values exist in columns that will not affect our analysis later on. All the information needed, will be extracted from the review_score column and thats because we will not get into further data wrangling with these columns. So now, we can change the data type of the last two columns because they represent a timestamp

In [48]:
order_reviews["review_creation_date"] = pd.to_datetime(order_reviews["review_creation_date"])

order_reviews["review_answer_timestamp"] =  pd.to_datetime(order_reviews["review_answer_timestamp"])

In [49]:
# Confirmation
order_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  object        
 1   order_id                 99224 non-null  object        
 2   review_score             99224 non-null  int64         
 3   review_comment_title     11568 non-null  object        
 4   review_comment_message   40977 non-null  object        
 5   review_creation_date     99224 non-null  datetime64[ns]
 6   review_answer_timestamp  99224 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(4)
memory usage: 5.3+ MB


In [50]:
# Duplicates
order_reviews.duplicated().sum()

0

In [51]:
order_reviews['review_id'].is_unique

False

In [52]:
order_reviews['order_id'].is_unique

False

Interesting! So review_id should be the id of each review. Order_id should be the id of each order. But none of them are unique which is kind of strange considering we dont have duplicate values. We shall dive deeper into this because maybe there is a mixup in the definitions of review_id and order_id I just gave

Why in the order_review dataset, review id hasnt unique values?

In [53]:
order_reviews[order_reviews.duplicated(subset='review_id', keep=False)].sort_values(by='review_id')

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
46678,00130cbe1f9d422698c812ed8ded1919,dfcdfc43867d1c1381bfaf62d6b9c195,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07,2018-03-20 18:08:23
29841,00130cbe1f9d422698c812ed8ded1919,04a28263e085d399c97ae49e0b477efa,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07,2018-03-20 18:08:23
90677,0115633a9c298b6a98bcbe4eee75345f,78a4201f58af3463bdab842eea4bc801,5,NaN,NaN,2017-09-21,2017-09-26 03:27:47
63193,0115633a9c298b6a98bcbe4eee75345f,0c9850b2c179c1ef60d2855e2751d1fa,5,NaN,NaN,2017-09-21,2017-09-26 03:27:47
92876,0174caf0ee5964646040cd94e15ac95e,f93a732712407c02dce5dd5088d0f47b,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07,2018-03-08 03:00:53
...,...,...,...,...,...,...,...
31120,fe5c833752953fed3209646f1f63b53c,4863e15fa53273cc7219c58f5ffda4fb,1,NaN,"Comprei dois produtos e ambos, mesmo enviados ...",2018-02-28,2018-02-28 13:57:52
7870,ff2fc9e68f8aabfbe18d710b83aabd30,2da58e0a7dcfa4ce1e00fad9d03ca3b5,2,NaN,NaN,2018-03-17,2018-03-19 11:44:15
82521,ff2fc9e68f8aabfbe18d710b83aabd30,1078d496cc6ab9a8e6f2be77abf5091b,2,NaN,NaN,2018-03-17,2018-03-19 11:44:15
73951,ffb8cff872a625632ac983eb1f88843c,c44883fc2529b4aa03ca90e7e09d95b6,3,NaN,NaN,2017-07-22,2017-07-26 13:41:07


Interesting! For the same review_id we have the exact same columns, except order_id. Lets see if the same thing is happening for order_id

In [54]:
order_reviews[order_reviews.duplicated(subset='order_id', keep=False)].sort_values(by='order_id')

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
25612,89a02c45c340aeeb1354a24e7d4b2c1e,0035246a40f520710769010f752e7507,5,NaN,NaN,2017-08-29,2017-08-30 01:59:12
22423,2a74b0559eb58fc1ff842ecc999594cb,0035246a40f520710769010f752e7507,5,NaN,Estou acostumada a comprar produtos pelo barat...,2017-08-25,2017-08-29 21:45:57
22779,ab30810c29da5da8045216f0f62652a2,013056cfe49763c6f66bda03396c5ee3,5,NaN,NaN,2018-02-22,2018-02-23 12:12:30
68633,73413b847f63e02bc752b364f6d05ee9,013056cfe49763c6f66bda03396c5ee3,4,NaN,NaN,2018-03-04,2018-03-05 17:02:00
854,830636803620cdf8b6ffaf1b2f6e92b2,0176a6846bcb3b0d3aa3116a9a768597,5,NaN,NaN,2017-12-30,2018-01-02 10:54:06
...,...,...,...,...,...,...,...
27465,5e78482ee783451be6026e5cf0c72de1,ff763b73e473d03c321bcd5a053316e8,3,NaN,Não sei que haverá acontecido os demais chegaram,2017-11-18,2017-11-18 09:02:48
41355,39de8ad3a1a494fc68cc2d5382f052f4,ff850ba359507b996e8b2fbb26df8d03,5,NaN,Envio rapido... Produto 100%,2017-08-16,2017-08-17 11:56:55
18783,80f25f32c00540d49d57796fb6658535,ff850ba359507b996e8b2fbb26df8d03,5,NaN,"Envio rapido, produto conforme descrito no anu...",2017-08-22,2017-08-25 11:40:22
92230,870d856a4873d3a67252b0c51d79b950,ffaabba06c9d293a3c614e0515ddbabc,3,NaN,NaN,2017-12-20,2017-12-20 18:50:16


Not exactly! Here all the other columns between the rows that have the same order_id are different. Different review_score, review_comment_message etc.

## **Interpretation**
Initial checks showed that neither review_id nor order_id is unique.


*   Duplicate review_id records were inspected using:
order_reviews[order_reviews.duplicated(subset='review_id', keep=False)]

These records shared identical review content, score, and timestamps, while being associated with different order_id values.

*   Duplicate order_id records were inspected using:
order_reviews[order_reviews.duplicated(subset='order_id', keep=False)]

These records contained different review_id values and often different review content and timestamps.

This is interesting. In the first case, it seems like a single review may be linked to multiple orders. In the second case, one order receives multiple reviews. This is more familiar, because someone can change the review after buying one product. For example, at the first glance, one gives a five star review based on the delivery time, but then after using the product, doesnt feel satisfied and changes the review to a 3 star review. One thing is for sure. Every row of this table is an association between an order and a review record.

## **Conclusion**
The repeated IDs were not treated as duplicate rows because the records contain meaningful relationship information and different combinations of order_id and review_id. No rows were removed during this step


## **Order_items dataset**

In [55]:
order_items.shape

(112650, 7)

In [56]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [57]:
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [58]:
# Turning shipping_limit_date to a timestamp
order_items["shipping_limit_date"] = pd.to_datetime(order_items["shipping_limit_date"])

In [59]:
order_items["order_id"].is_unique

False

In [60]:
order_items["product_id"].is_unique

False

This is expected of course. The same order can have two or more products. This means that each row contains the information of the quantity of the product, the seller of the product etc. This is clear. Obviously product_id, or seller_id couldnt act as a primary key here, because in seperate orders we can have the same product, or the same seller.

In [61]:
order_items.duplicated().sum()

0

## **Orders dataset**

In [62]:
orders.shape

(99441, 8)

In [63]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [64]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


We detect missing values at 3 different columns. Also the last 5 columns have the wrong data type. I will do that first and we will see later how I will handle the missing values.  

In [65]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

orders[date_columns] = orders[date_columns].apply(pd.to_datetime)

In [66]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

The absence of the values could occur to the fact that not all deliveries were delivered.

In [67]:
orders.groupby("order_status")[
    "order_delivered_customer_date"
].agg(
    total="size",
    missing=lambda x: x.isna().sum()
)

,total,missing
order_status,,
approved,2,2
canceled,625,619
created,5,5
delivered,96478,8
invoiced,314,314
processing,301,301
shipped,1107,1107
unavailable,609,609


This hypothesis seems correct according to the above table. For the order_delivered_customer_date column, the most missing values come from orders that have cancelled, shipped or unavailable on their status, which make total sense. The strange thing here is that 6 orders that were cancelled seem to have a delivery date, which is interesting to look further into. Also, 8 orders that were deliverd doesnt have the date that they were delivered. Besides from that, everything looks the way they should be.

In [68]:
orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_customer_date"].isna())
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaT,2017-12-18
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaT,2018-07-16
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaT,2018-07-30
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaT,2018-07-30
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaT,2018-07-24
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaT,2018-06-26
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaT,2018-07-19


Among the 96,478 orders with status "delivered", only 8 records are missing the customer delivery timestamp. Most of these orders still contain valid approval and carrier timestamps, suggesting that the orders were likely completed but the final delivery timestamp was not recorded. One record is missing both carrier and customer delivery timestamps despite being marked as delivered, indicating a potential data quality issue. Because the inconsistent values are too small of a quantity compared to the rest of the data, the values will stay NaN. If further analysis occurs, it will be considered.

In [69]:
orders[
    (orders["order_status"] == "canceled") &
    (orders["order_delivered_customer_date"].notna())
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
2921,1950d777989f6a877539f53795b4c3c3,1bccb206de9f0f25adc6871a1bcf77b2,canceled,2018-02-19 19:48:52,2018-02-19 20:56:05,2018-02-20 19:57:13,2018-03-21 22:03:51,2018-03-09
8791,dabf2b0e35b423f94618bf965fcb7514,5cdec0bb8cbdf53ffc8fdc212cd247c6,canceled,2016-10-09 00:56:52,2016-10-09 13:36:58,2016-10-13 13:36:59,2016-10-16 14:36:59,2016-11-30
58266,770d331c84e5b214bd9dc70a10b829d0,6c57e6119369185e575b36712766b0ef,canceled,2016-10-07 14:52:30,2016-10-07 15:07:10,2016-10-11 15:07:11,2016-10-14 15:07:11,2016-11-29
59332,8beb59392e21af5eb9547ae1a9938d06,bf609b5741f71697f65ce3852c5d2623,canceled,2016-10-08 20:17:50,2016-10-09 14:34:30,2016-10-14 22:45:26,2016-10-19 18:47:43,2016-11-30
92636,65d1e226dfaeb8cdc42f665422522d14,70fc57eeae292675927697fe03ad3ff5,canceled,2016-10-03 21:01:41,2016-10-04 10:18:57,2016-10-25 12:14:28,2016-11-08 10:58:34,2016-11-25
94399,2c45c33d2f9cb8ff8b1c86cc28c11c30,de4caa97afa80c8eeac2ff4c8da5b72e,canceled,2016-10-09 15:39:56,2016-10-10 10:40:49,2016-10-14 10:40:50,2016-11-09 14:53:50,2016-12-08


This might represent orders that were deliverd but later customers asked for refunds or just inconsistent data. Based on the lack of information I will those values as they are.

In [70]:
orders["order_id"].is_unique

True

Based on the format of the table, this was expected. We can see order_id acting like the primary key of the table, which means that every row in the table represents an order that was placed by a customer

In [71]:
orders.duplicated().sum()

0

## **Products Dataset**

In [72]:
products.shape

(32951, 9)

In [73]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [74]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [75]:
products.duplicated().sum()

0

In [76]:
products["product_id"].is_unique

True

Obviously here product_id acts like the primary key, which means that each row of the table represents one product and the rest of the columns contain inforamtion about that specific product. With that being said, I think that there is no reason for me to overanalyse this dataset just because it will not provide me with some exciting info or result that is worth being investigated. Later on, I will change the product_category_name column to english to be more understandable to the public. Moving on

##**Product category Name Translation Dataset**

In [77]:
product_category_name_translation.shape

(71, 2)

In [78]:
product_category_name_translation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [79]:
product_category_name_translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [80]:
product_category_name_translation.duplicated().sum()

0

This is the translation I was talking about earlier. So now i want to change the names of the products from the pruducts dataset to english

In [81]:
translation_dict = dict(
    zip(
        product_category_name_translation["product_category_name"],
        product_category_name_translation["product_category_name_english"]
    )
)

products["product_category_name"] = (
    products["product_category_name"]
    .map(translation_dict)
)

In [82]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumery,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,art,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,sports_leisure,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,baby,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,housewares,37.0,402.0,4.0,625.0,20.0,17.0,13.0


## **Order Payment Dataset**

In [83]:
order_payment.shape

(103886, 5)

In [84]:
order_payment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [85]:
order_payment.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [86]:
order_payment['order_id'].is_unique

False

In [87]:
order_payment["payment_sequential"].value_counts()

payment_sequential
1     99360
2      3039
3       581
4       278
5       170
6       118
7        82
8        54
9        43
10       34
11       29
12       21
13       13
14       10
15        8
18        6
19        6
16        6
17        6
21        4
20        4
22        3
26        2
24        2
23        2
25        2
29        1
28        1
27        1
Name: count, dtype: int64

In [88]:
order_payment["order_id"].nunique()

99440

In [89]:
orders.shape[0]

99441

In [90]:
missing_payment_orders = (
    set(orders["order_id"]) -
    set(order_payment["order_id"])
)

missing_payment_orders

{'bfbd0f9bdef84302105ad712db648a6c'}

In [91]:
orders[orders["order_id"].isin(missing_payment_orders)]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
30710,bfbd0f9bdef84302105ad712db648a6c,86dc2ffce2dfff336de2f386a786e574,delivered,2016-09-15 12:16:38,2016-09-15 12:16:38,2016-11-07 17:11:53,2016-11-09 07:47:38,2016-10-04


***Data Integrity Check – Orders vs. Payments***

The orders dataset contains 99,441 unique orders, while the order_payments dataset contains payment records for 99,440 unique orders. A comparison of the primary keys revealed one delivered order (bfbd0f9bdef84302105ad712db648a6c) with no corresponding payment record. Since this represents a single observation (≈0.001% of all orders), it was documented as a data quality inconsistency and retained in the dataset.

In [92]:
order_payment["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

In [ ]:
# Lastly we check for duplicates
order_payment.duplicated().sum()

Now we export our clean data and save them into our folders

In [99]:
customers.to_csv("../data/processed/customers_clean.csv", index=False)
orders.to_csv("../data/processed/orders_clean.csv", index=False)
geolocation.to_csv("../data/processed/geolocation_clean.csv", index=False)
order_items.to_csv("../data/processed/order_items_clean.csv", index=False)
order_payment.to_csv("../data/processed/order_payments_clean.csv", index=False)
order_reviews.to_csv("../data/processed/order_reviews_clean.csv", index=False)
products.to_csv("../data/processed/products_clean.csv", index=False)
sellers.to_csv("../data/processed/sellers_clean.csv", index=False)